### ISSUE TITLE
Chain Reductions pattern matching table

### ISSUE THREAD

**Issue by tom-huntington** [codereport/jello#36]
I wrote a table to help you know how to "peel" items of the chains (perform reductions). In a monadic or dyadic chain, reductions always result in an monad or dyad respectively. (Nilads can just be promoted to constant n-ads in an n-adic chain???)

<table>
    <tr>
        <td>

| chain |  dyadic |
| :------------------------:| :----: |
| `2I 2 {1…,} `| `ε" {1…,}` |
| `2 2 {1…,} `| `ε' {1…,}` |
| `2 2 2 …` | `Φ₁ …` |
| `2 1 …`   | `B₁ …` |
| `1 2 {1…,}`|  `Δ {1…,}` |
| `1 2 2 …` | `Φ.₂ …` |
| `1 1 …`  | `B.₅ …` |

</td>
        <td>

| chain      | monadic |
| :------:   | :-----: |
| `1 2 1 …`  | `Φ …` |
| `1 2 {2…,}`| `Σ {2…,}` |
| `1 1 …`    | `B …` |
| `2 {2…,}`  | `W {2…,}` |
| `2 1 …`    | `S …` |

</td>
    </tr>
</table>

```
ε" f g = x y -> f(x, g(x,y))
ε' f g = x y -> g(f(x,y), y)
B.₅ f g = x y -> g(f(x)) 
```

I found some issues with Jello's combinator trees:

- [ ] Same issue https://github.com/codereport/jello/issues/14 for B₁ as we had for Φ₁
- [ ] ε' isn't the same combinator depending on the whether its an initial combinator or not, but Jello doesn't distinguish. 

ε' issue
--

 The pairs here associate to the right ` [1, [1, 10]]`
``` 
> 1 10 :: pair pair
           ,    ,
   ,  1 10 ➡️  [1, 10]
   ,, 1 10 ➡️  [1, [1, 10]]
    This is a 2-2 dyadic chain (ε')
              └┬┘
              ε'
```

but the pairs here associate to the left `[[2, 11], 10]`
```
> 1 10 :: pair add1 pair
           ,    ‘    ,
   ,   1 10 ➡️  [1, 10]
   ,‘  1 10 ➡️  [2, 11]
   ,‘, 1 10 ➡️  [[2, 11], 10]
    This is a 2-1-2 dyadic chain (B₁ε')
              └┬┘ ⋮
              B₁  ⋮
               └┬─┘
               ε'
```
although 2-1-2 reduces to 2-2 so the only difference is the first dyad in one is initial, and the other is a dyadic chain.

**Comment by tom-huntington** [comment]
<table>
    <tr>
        <td>



| chain |  dyadic |
| :------------------------:| :----: |
| `2 2 2 …` | `Φ₁ …` |
| `2 2 … `| `ε{",'} …`* |
| `2 1 …`   | `B₁ …` |
| `1 2 2 …` | `Φ.₂ …` |
| `1 2 …`|  `Δ …` |
| `1 1 …`  | `B.₅ …` |


</td>
        <td>


| chain      | monadic |
| :------:   | :-----: |
| `1 2 1 …`  | `Φ …` |
| `1 2 …`| `Σ …` |
| `1 1 …`    | `B …` |
| `2 1 …`    | `S …` |
| `2 …`  | `W …` |




</td>
    </tr>
</table>

Perhaps the table could be written this way, provided it is clear that we are matching the most specialized match first.

**Comment by codereport** [comment]
Thanks for this! Will address all in the future. 

I also want to change Jellyfish to work such that the first chain in a multichain has a fixed arity (I think that is what I want at least). Added these spellings recently ⬇️ but when I modified Jellyfish do have [better support for D2](https://github.com/codereport/jellyfish/commit/f9c6cb490c585c117d54609b646f15aeb048291e) I couldn't figure out how to register the first chain of a multichain as monadic when it is called in a dyadic context. If you have thoughts let me know : )

![image](https://github.com/codereport/jello/assets/36027403/97a3007e-4529-43a9-bfe9-f27a6545a821)

**Comment by tom-huntington** [comment]
@codereport why can you just start with a monadic separator to specify a subchain
```
> 1 10 :: . pair : pair
          µ  ,   ð  ,
   µ,   1 10 ➡️  [1, 1]
   µ,ð, 1 10 ➡️  [[1, 1], 10]
    This is a s-2-s-2 dyadic chain (WB)
                │   │
                W   d
                └─┬─┘
                  B
```

I kind of think your mental model is a bit confused on how the reductions happen. The arity of the whole chain is based on the arguments/context. What you are actually thinking of as the stuff after "1st chain" is actually a subchain that reduces down into a monad.

A less confusing way to denote this is to explicitly use parentheses `(` for a monadic sub chain, curly parentheses {` for a dyadic subchain.
```
< . sort rev sums . sum . sum add1
{{< (sort rev sums) sum} (sum add1)}
```
`sum add1` is not the rest of the chain, it is a subchain that reduces down to a monad. That monadic chain can refer back to the original argument, which is helpful, but you can add a separator to end the subchain and continue the chain.


Breaking change
===

Also are you aware that your [better support for D2](https://github.com/codereport/jellyfish/commit/f9c6cb490c585c117d54609b646f15aeb048291e) is a breaking change

`2 1 1` use to be `B₁B₁` now it is D₂.

**Comment by tom-huntington** [comment]
https://github.com/DennisMitchell/jellylanguage/wiki/Tutorial
> Multi-chain links let us do just that! To construct them, we use the chain separators `øµðɓ`

Surely one of these will let you end a subchain without starting another

**Comment by codereport** [comment]
My mental model is/was definitely wrong. I didn't realize the chain is formed and then the context still matters. I also feel like i don't totally understand explicitly what the delineation of link/subchain/chain is.

How did you learn all of this? Just by trial and error? or looking at the source code?

**Comment by tom-huntington** [comment]
@codereport I'm trained as a mathematician, it's just the logical way to do these operations. Denis implemented it quite different, more like a stack based language. So my mental model is a bit different to his implementation

**Comment by codereport** [comment]
Interesting. And yea, I realized that I could have just used a monadic separator to get what I wanted with the d2 combinator. But I like the expressions having a fixed meaning i.e. it can only be called monadically or dyadically. Maybe after I play around with it maybe I wouldn't like it as much though. You might be right that the subchain ender is what I want.

**Comment by tom-huntington** [comment]
@codereport I agree that you should have to specify the arity at function definition, and throw an error if the wrong number of arguments are passed. However, your d2 change is orthogonal.

I think you got tripped up by the fact that Denis' implementation is using the fact the the following are equivalent:
```
f₁ g₂ h₁ k₂ : ...
. f₁ : g₂ . h₁ : k₂ : ...
```
IMO, It's better to think of jelly working as the second, each atom is implicitly separated until the first separator.

Also my mental model: jelly is a language for writing extremely nested compositions of functions. At the end of the day, it should be implemented by construct a single function then calling it, rather than Denis' the stack based implementation.
